lua/neotest-kotlin/init.lua

```diff
@@ -30,84 +30,100 @@ end
 ---@async
 ---@param file_path string Absolute file path
 ---@return neotest.Tree | nil
 function M.Adapter.discover_positions(file_path)
-  local positions = lib.treesitter.parse_positions(file_path, treesitter_query, {
-    nested_namespaces = true,
-    nested_tests = false,
-  })
+	local positions = lib.treesitter.parse_positions(file_path, treesitter_query, {
+		nested_namespaces = true,
+		nested_tests = false,
+	})
 
-  return positions
+	return positions
 end
 
+---Determines the package of a directory
+---@param dir string
+---@return string? package
+local function dir_determine_package(dir)
+	if not lib.files.is_dir(dir) then
+		error(string.format("expected '%s' be a directory, but it's not", dir))
+	end
+
+	local test_file = nil
+	local files = vim.fn.globpath(dir, "**/*.kt", false, true)
+	for _, file in ipairs(files) do
+		if filter.is_test_file(file) then
+			test_file = file
+			break
+		end
+	end
+
+	if test_file == nil then
+		return nil
+	end
+
+	return position_parser.get_first_match_string(test_file, package_query)
+end
+
+---@class Context
+---@field results_path string path to the results file
+---@field path string path to the directory/file
+
+---@class neotest.RunSpec
+---@field cwd string?
+---@field context Context
+---@field command string
+
 ---@param args neotest.RunArgs
 ---@return nil | neotest.RunSpec | neotest.RunSpec[]
 function M.Adapter.build_spec(args)
-  local results_path = async.fn.tempname() .. ".json"
-
-  -- Write something so there is a place to stream to...
-  lib.files.write(results_path, "")
-
-  local tree = args.tree
-
-  if not tree then
-    return
-  end
-
-  local pos = tree:data()
-
-  local root = M.Adapter.root(pos.path)
-  local pkg = position_parser.get_first_match_string(pos.path, package_query)
-  local className = position_parser.get_first_match_string(pos.path, class_query)
-  local specPackage = pkg .. "." .. className
-  local tests = "*"
-
-  local gradle_command = command.parse(tests, specPackage, results_path)
-
-  local stream_data, stop_stream = lib.files.stream_lines(results_path)
-
-  print("command: " .. gradle_command)
-
-  local all_results = {}
-
-  return {
-    command = gradle_command,
-    cwd = root,
-    context = {
-      all_results = all_results,
-      results_path = results_path,
-      file = pos.path,
-      stop_stream = stop_stream,
-    },
-    stream = function()
-      return function()
-        local new_results = stream_data()
-        local success, parsed_result = pcall(output_parser.parse_lines, new_results, pos.path, specPackage)
-        if not success then
-          print("An error ocurred while attempting to stream data to result: " ..
-            vim.inspect(err) .. " new_results: " .. vim.inspect(new_results))
-          return nil
-        else
-          -- merge the parsed results with all results...
-          for k, v in pairs(parsed_result) do all_results[k] = v end
-          return parsed_result
-        end
-      end
```

Comment:
Don't really understand the benefit of streaming here, since we need to read the entirety of the output. The main benefit of streaming from the docs seems to be that you can easily read the output if all you're doing is appending to the file, but since we create a new file every time it doesn't really do anything differently than just read everything

Reply:
I think at the time this may have been the only way to do this in neotest. 

Alternatively and very likely, I thought that it was the only way to do this in neotest, and was mistaken 😄

Reply:
I don't know which is the best route, but I think parsing the whole file in the async results function feels simpler? Can always pivot if we find performance issues or something 👍